File: heartwood-mcp/tools/documents.py
"""Document tools — invoices, estimates, purchase orders, etc."""

from __future__ import annotations

from app import mcp
from constants import ORG_ID
from pave import call_pave, flatten_nodes


@mcp.tool()
async def jt_create_document(
    job_id: str,
    document_type: str,
    account_id: str,
    template_id: str | None = None,
    name: str | None = None,
) -> dict:
    """Create a document (invoice, estimate, purchase order, etc.) on a job.

    Args:
        job_id: The job ID.
        document_type: Type of document: "invoice", "estimate", "purchaseOrder", "changeOrder".
        account_id: The account (customer/vendor) this document is for.
        template_id: Optional template ID to use.
        name: Optional document name/title.
    """
    params: dict = {
        "organizationId": ORG_ID,
        "jobId": job_id,
        "type": document_type,
        "accountId": account_id,
    }
    if template_id:
        params["templateId"] = template_id
    if name:
        params["name"] = name

    return await call_pave(
        "createDocument",
        params=params,
        fields={"id": {}, "number": {}, "name": {}, "type": {}, "status": {}},
    )


@mcp.tool()
async def jt_update_document(
    document_id: str,
    status: str | None = None,
    name: str | None = None,
) -> dict:
    """Update a document's status or name.

    Args:
        document_id: The document ID.
        status: New status (e.g. "draft", "sent", "approved", "void").
        name: New name.
    """
    params: dict = {"id": document_id}
    if status is not None:
        params["status"] = status
    if name is not None:
        params["name"] = name

    return await call_pave(
        "updateDocument",
        params=params,
        fields={"id": {}, "name": {}, "status": {}},
    )


@mcp.tool()
async def jt_get_documents(
    job_id: str | None = None,
    document_type: str | None = None,
    status: str | None = None,
    first: int = 50,
) -> list[dict]:
    """List documents, optionally filtered by job, type, or status.

    Args:
        job_id: Filter by job.
        document_type: Filter by type ("invoice", "estimate", etc.).
        status: Filter by status.
        first: Max results.
    """
    filter_obj: dict = {}
    if job_id:
        filter_obj["jobId"] = {"eq": job_id}
    if document_type:
        filter_obj["type"] = {"eq": document_type}
    if status:
        filter_obj["status"] = {"eq": status}

    params: dict = {"first": first}
    if filter_obj:
        params["filter"] = filter_obj

    result = await call_pave(
        "documents",
        params=params,
        fields={
            "nodes": {
                "id": {},
                "number": {},
                "name": {},
                "type": {},
                "status": {},
                "jobId": {},
                "accountId": {},
                "total": {},
                "createdAt": {},
            }
        },
    )
    return flatten_nodes(result)


@mcp.tool()
async def jt_get_document_line_items(
    document_id: str,
    first: int = 200,
) -> list[dict]:
    """Get line items for a specific document.

    Args:
        document_id: The document ID.
        first: Max results.
    """
    result = await call_pave(
        "documentLineItems",
        params={"filter": {"documentId": {"eq": document_id}}, "first": first},
        fields={
            "nodes": {
                "id": {},
                "name": {},
                "description": {},
                "quantity": {},
                "unitPrice": {},
                "amount": {},
                "groupName": {},
            }
        },
    )
    return flatten_nodes(result)


@mcp.tool()
async def jt_get_document_templates(
    document_type: str | None = None,
    first: int = 50,
) -> list[dict]:
    """Get available document templates.

    Args:
        document_type: Filter by type ("invoice", "estimate", etc.).
        first: Max results.
    """
    filter_obj: dict = {"organizationId": {"eq": ORG_ID}}
    if document_type:
        filter_obj["type"] = {"eq": document_type}

    result = await call_pave(
        "documentTemplates",
        params={"filter": filter_obj, "first": first},
        fields={
            "nodes": {
                "id": {},
                "name": {},
                "type": {},
            }
        },
    )
    return flatten_nodes(result)

File: workspace/projects/heartwood-mcp/src/tools/jt/documents.ts
/**
 * JT Document tools — 5 tools
 */

import type { PaveClient, ToolResult } from "../../pave/index.js";
import { DOCUMENT_FIELDS, DOCUMENT_LINE_ITEM_FIELDS, TEMPLATE_FIELDS } from "../../pave/index.js";
import type { ToolDef } from "../registry.js";

export function documentTools(pave: PaveClient): ToolDef[] {
  return [
    // ── jt_create_document ─────────────────────────────────────────────
    {
      name: "jt_create_document",
      description:
        "Create a document (customerOrder, customerInvoice, vendorOrder, vendorBill, bidRequest).",
      inputSchema: {
        type: "object" as const,
        properties: {
          jobId: { type: "string", description: "Job ID" },
          type: {
            type: "string",
            enum: ["customerOrder", "customerInvoice", "vendorOrder", "vendorBill", "bidRequest"],
            description: "Document type",
          },
          accountId: { type: "string", description: "Account ID (optional)" },
          documentTemplateId: { type: "string", description: "Template ID (optional)" },
          costCodeIds: {
            type: "array", items: { type: "string" },
            description: "Cost code IDs to include (optional)",
          },
          costGroupNames: {
            type: "array", items: { type: "string" },
            description: "Cost group names to include (optional)",
          },
          costItemIds: {
            type: "array", items: { type: "string" },
            description: "Cost item IDs to include (optional)",
          },
          costItemOverrides: {
            type: "array",
            items: {
              type: "object",
              properties: {
                costItemId: { type: "string" },
                quantity: { type: "number" },
                unitPrice: { type: "number" },
              },
            },
            description: "Cost item overrides (optional)",
          },
          date: { type: "string", description: "Document date (optional)" },
          name: { type: "string", description: "Document name (optional)" },
          subject: { type: "string", description: "Subject line (optional)" },
          description: { type: "string", description: "Description (optional)" },
          footer: { type: "string", description: "Footer text (optional)" },
          externalId: { type: "string", description: "External ID (optional)" },
          taxRate: { type: "number", description: "Tax rate as decimal (optional)" },
        },
        required: ["jobId", "type"],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        const data: Record<string, unknown> = {
          jobId: params.jobId,
          type: params.type,
        };
        const optionalFields = [
          "accountId", "documentTemplateId", "costCodeIds", "costGroupNames",
          "costItemIds", "costItemOverrides", "date", "name", "subject",
          "description", "footer", "externalId", "taxRate",
        ];
        for (const field of optionalFields) {
          if (params[field] !== undefined) data[field] = params[field];
        }
        return pave.create("document", data, DOCUMENT_FIELDS);
      },
    },

    // ── jt_update_document ─────────────────────────────────────────────
    {
      name: "jt_update_document",
      description: "Update a document's status, description, cost items, or push to QBO.",
      inputSchema: {
        type: "object" as const,
        properties: {
          documentId: { type: "string", description: "Document ID" },
          status: { type: "string", description: "New status (optional)" },
          description: { type: "string", description: "New description (optional)" },
          costItemUpdates: {
            type: "array",
            items: {
              type: "object",
              properties: {
                costItemId: { type: "string" },
                quantity: { type: "number" },
                unitPrice: { type: "number" },
              },
            },
            description: "Cost item updates (optional)",
          },
          pushToQbo: { type: "boolean", description: "Push to QuickBooks Online (optional)" },
        },
        required: ["documentId"],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        const data: Record<string, unknown> = {};
        if (params.status) data.status = params.status;
        if (params.description) data.description = params.description;
        if (params.costItemUpdates) data.costItemUpdates = params.costItemUpdates;
        if (params.pushToQbo !== undefined) data.pushToQbo = params.pushToQbo;
        return pave.update("document", params.documentId as string, data, DOCUMENT_FIELDS);
      },
    },

    // ── jt_get_documents ───────────────────────────────────────────────
    {
      name: "jt_get_documents",
      description: "Get documents, optionally filtered by job and/or type.",
      inputSchema: {
        type: "object" as const,
        properties: {
          jobId: { type: "string", description: "Filter by job ID (optional)" },
          type: {
            type: "string",
            enum: ["customerOrder", "customerInvoice", "vendorOrder", "vendorBill", "bidRequest"],
            description: "Filter by document type (optional)",
          },
        },
        required: [],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        const conditions: Array<{ field: string; operator: string; value: unknown }> = [];
        if (params.jobId) conditions.push({ field: "jobId", operator: "eq", value: params.jobId });
        if (params.type) conditions.push({ field: "type", operator: "eq", value: params.type });
        return pave.query({
          entity: "document",
          fields: DOCUMENT_FIELDS,
          filter: conditions.length > 0 ? { operator: "and", conditions } : undefined,
        });
      },
    },

    // ── jt_get_document_line_items ─────────────────────────────────────
    {
      name: "jt_get_document_line_items",
      description: "Get the line items (groups + items) for a specific document.",
      inputSchema: {
        type: "object" as const,
        properties: {
          documentId: { type: "string", description: "Document ID" },
        },
        required: ["documentId"],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        return pave.query({
          entity: "documentLineItem",
          fields: DOCUMENT_LINE_ITEM_FIELDS,
          filter: { conditions: [{ field: "documentId", operator: "eq", value: params.documentId }] },
        });
      },
    },

    // ── jt_get_document_templates ──────────────────────────────────────
    {
      name: "jt_get_document_templates",
      description: "Get document templates by type. Use to find template ID before creating a document.",
      inputSchema: {
        type: "object" as const,
        properties: {
          type: {
            type: "string",
            enum: ["customerOrder", "customerInvoice", "vendorOrder", "vendorBill", "bidRequest"],
            description: "Document type to get templates for",
          },
        },
        required: ["type"],
      },
      handler: async (params: Record<string, unknown>): Promise<ToolResult> => {
        return pave.query({
          entity: "documentTemplate",
          fields: TEMPLATE_FIELDS,
          filter: { conditions: [{ field: "type", operator: "eq", value: params.type }] },
        });
      },
    },
  ];
}
